tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,1279 @@
+
+	-- store the error function independent of the environment
+	local error = error
+	exit = {
+		error = function() error("exited with error") end,
+		warn  = function() error("exited with warn") end,
+		succ  = function() error("exited with succ") end,
+	}
+else
+	local os_exit = os.exit
+	exit = {
+		error = function() os_exit(11) end,
+		warn  = function() os_exit(10) end,
+		succ  = function() os_exit(0) end,
+	}
+end
+
+do
+	-- I don't like using the debug library, but getting a traceback here is a
+	-- must to find where the error originates from
+	local debug_traceback = debug.traceback
+	-- use the lua error function -> exits immediately
+	local error = error
+
+	-- Prevent trying to change the environment.
+	local function bad_index(...)
+		local msg = "Attempt to access an undefined index:"
+		for i = 2, select("#", ...) do
+			msg = msg ..tostring(select(i, ...)).." "
+		end
+		msg = msg.."\n\n"..debug_traceback(nil, 2)
+		error(msg)
+	end
+	setmetatable(env, {
+		__index     = bad_index,
+		__metatable = false,
+		__newindex  = bad_index,
+	})
+end
+
+_ENV = env
+----------------------------------
+-- restricted area startes here --
+----------------------------------
+
+---------------
+-- pathutils --
+---------------
+-- -> probably moved to a different library eventually
+local pathlib = {}
+do
+	-- other projects like penlight or l3build do stuff with different pathseps, according to
+	-- https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#canonicalize-separators
+	-- and
+	-- https://retrocomputing.stackexchange.com/questions/28344/since-when-does-windows-support-forward-slash-as-path-separator
+	-- windows since quite a while also works with / as pathsep.
+	-- Thus, this pathlib will normalize paths for having / as pathsep before working on paths
+	if os.type == "windows" then
+		---Normalize path such that windows also uses /
+		---@param path string
+		---@return string
+		function pathlib.path_normalize(path)
+			return path:gsub("\\", "/")
+		end
+	else
+		---Unix already uses / as pathsep -> no-op
+		---@param path string
+		---@return string
+		function pathlib.path_normalize(path)
+			return path
+		end
+	end
+
+	-- still windows paths work with disk specifiers -> special handling required
+	if os.type == "windows" then
+		---Check if path is an absolute path
+		---@param path string
+		---@return boolean? is_abs
+		---@return string? err_msg
+		function pathlib.path_is_absolute(path)
+			local err
+			path, err = pathlib.sanitize_path(path)
+			if not path then return nil, err end
+
+			return path:sub(2,2) == ":" and path:sub(3,3) == "/"
+		end
+	else
+		---Check if path is an absolute path
+		---@param path string
+		---@return boolean? is_abs
+		---@return string? err_msg
+		function pathlib.path_is_absolute(path)
+			local err
+			path, err = pathlib.sanitize_path(path)
+			if not path then return nil, err end
+
+			return path:match("^/") and true or false
+		end
+	end
```

Comment:
`pathlib.path_is_absolute("/home/max/../../etc/passwd")` and `pathlib.path_is_absolute("///home///max///texfile.tex")` would both return `true`, even though I wouldn't consider those to be absolute paths. The following test should catch those:

```lua
if path:match("/%.*/")
    return false
end
```

All of this should be caught by `kpse.out_name_ok_silent_extended` though, so I don't think that there's any security concern with this.

Reply:
I agree, still it's nice to catch that case. The check looks good, I'd just be a bit less restrictive and still allow `pathlib.path_is_absolute("///home///max///texfile.tex")` as technically it's still a valid and absolute path. The other example is a clear security concern (if `kpse` doesn't catch it).

So I'd go with the following pattern `/%.%.+/` (which also allows `/home/./max/texfile.tex` as this is still not a security issue and might happen in some cases of path construction).

Can you see any trouble in this pattern? (in this regard it's always nice to have a second mind think about it)